### PR TITLE
Update libovsdbops to use Where().List() instead of Get()

### PR DESCRIPTION
**Pull request by npinaeva**
to do index-based search.
Update the logic for model lookup to
1. Use index if any provided
2. If no indices provided with the Model, use predicate search.

Update syncPods find logical switch to only set Model.UUID if it is not
empty, otherwise search will only be performed using given UUID, and no
attempt to use predicate will be made.

Update egress_ip test to set empty ls UUID, since we don't know the real
switch UUID in the db to make a search, but it can be done by name.

Signed-off-by: Nadia Pinaeva <npinaeva@redhat.com>

### CODE REVIEW

**Comment by dcbw** [comment]
@npinaeva I presume there's a performance benefit here? IIRC get was slower than List() if we know the index already.

**Comment by npinaeva** [comment]
It is a preparation step to start using client indices, but I need to ask @jcaamano about how bad it will be for performance.
Potentially we can use get for db indices and list for client indices

**Comment by npinaeva** [comment]
/hold
until FF, September 16
but reviews are welcome :)

**Comment by jcaamano** [comment]
> @npinaeva I presume there's a performance benefit here? IIRC get was slower than List() if we know the index already.

@dcbw to use client indexes we need `Where`. It is still an indexed lookup in the cache. List() might run through a handful more lines of code but I would expect the difference ineligible unless I am missing something.

If we are concerned about this, we can use `Get` when we are provided with schema indexes and `Where` if not.

File: go-controller/pkg/libovsdbops/model_client.go [hunk 1]
```diff
@@ -386,52 +387,58 @@ func (m *modelClient) Lookup(opModels ...operationModel) error {
...
 	var err error
-	if opModel.Model != nil && !opModel.BulkOp {
-		err = m.get(opModel)
-		if err == nil || err != client.ErrNotFound {
+	if opModel.Model != nil {
+		err = m.list(opModel)
+		if err == nil && reflect.Indirect(reflect.ValueOf(opModel.ExistingResult)).Len() == 0 {
+			// index was provided, but nothing was found
+			return client.ErrNotFound
+		}
```
[the remark this answers is not in the document]
Comment (npinaeva) [reply go-controller/pkg/libovsdbops/model_client.go:402]: I tried to copy `get` behaviour when index is provided. `Get` will return `ErrNotFound` in that case, but I since it is ignored in `buildOps` https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/libovsdbops/model_client.go#L277 we can not return it here

File: go-controller/pkg/libovsdbops/model_client.go [hunk 2]
```diff
@@ -386,52 +387,58 @@ func (m *modelClient) Lookup(opModels ...operationModel) error {
...
 	}
-	uuid := getUUID(opModel.Model)
-	if uuid == "" || isNamedUUID(uuid) {
-		setUUID(opModel.Model, getUUID(copy))
+	if opModel.Model == nil || opModel.BulkOp {
+		return nil
 	}
-	return addToExistingResult(copy, opModel.ExistingResult)
+	// for non-bulk op cases, copy (the one) uuid found to model provided.
+	// it allows to use model UUID in DoAfter function to update references.
```
[the remark this answers is not in the document]
Comment (npinaeva) [reply go-controller/pkg/libovsdbops/model_client.go:435]: Yep agree, here is not a good place, I just don't like the fact that some `DoAfter` function use this side-effect of search optimization, I will try to add some comments in other places to explain that.

File: go-controller/pkg/ovn/egressip_test.go [hunk 3]
```diff
@@ -3523,8 +3523,8 @@ var _ = ginkgo.Describe("OVN master EgressIP Operations", func() {
 						Items: []v1.Pod{egressPod1},
 					},
 				)
-
-				fakeOvn.controller.lsManager.AddNode(node1.Name, node1.Name+"-UUID", []*net.IPNet{ovntest.MustParseIPNet(v4NodeSubnet)})
+				// we don't know the real switch UUID in the db, but it can be found by name
+				fakeOvn.controller.lsManager.AddNode(node1.Name, "", []*net.IPNet{ovntest.MustParseIPNet(v4NodeSubnet)})
```
[the remark this answers is not in the document]
Comment (npinaeva) [reply go-controller/pkg/ovn/egressip_test.go:3527]: I think real db UUIDs are generated, so if you get this LS from the db you will see something like 
`{UUID:83290196-e30a-4d93-a101-958d5c6fdfbb ACLs:[] Copp:<nil> DNSRecords:[] ExternalIDs:map[] ForwardingGroups:[] LoadBalancer:[] LoadBalancerGroup:[] Name:node1 OtherConfig:map[] Ports:[] QOSRules:[]}`
and the UUID you mentioned is used to check if db contents match expectations?

File: go-controller/pkg/ovn/pods.go [hunk 4]
```diff
@@ -103,7 +103,10 @@ func (oc *Controller) syncPodsRetriable(pods []interface{}) error {
 		sw := nbdb.LogicalSwitch{
 			Name: n.Name,
 		}
-		sw.UUID, _ = oc.lsManager.GetUUID(n.Name)
+		if swUUID, ok := oc.lsManager.GetUUID(n.Name); ok {
+			// do db search based on UUID if it is known
+			sw.UUID = swUUID
+		}
```
[the remark this answers is not in the document]
Comment (npinaeva) [reply go-controller/pkg/ovn/pods.go:109]: this change is not to fix test, but to make sure we don't set UUID for switches that are not known (actually empty string will be considered zeroIndex, and predicate search will be executed anyway, but I wanted to be a bit more strict here, and not rely on zero values comparison)
I will update this to only use non-empty UUIDs

File: go-controller/pkg/libovsdbops/model_client.go [hunk 5]
```diff
@@ -386,52 +387,58 @@ func (m *modelClient) Lookup(opModels ...operationModel) error {
...
 	}
-
+	// if index wasn't provided by the Model (errNoIndices) or Model == nil, try predicate search
 	if opModel.ModelPredicate != nil {
 		err = m.whereCache(opModel)
-	} else if opModel.BulkOp {
-		panic("Expected a ModelPredicate with BulkOp==true")
+	} else {
+		// list didn't find anything and ModelPredicate is not specified, return ErrNotFound
+		return client.ErrNotFound
```
[the remark this answers is not in the document]
Comment (npinaeva) [reply go-controller/pkg/libovsdbops/model_client.go:413]: Actually when creating an object, we don't specify neither index, nor predicate, and `bulkOps` is handled in `buildOps`, so I can return either nil or `ErrNotFound` here

[the remark this answers is not in the document]
Comment (jcaamano) [reply go-controller/pkg/libovsdbops/model_client.go]: How is `BulkOp` handled in `buildOps`?
This is detecting an usage error when `BulkOp==true`, `ModelPredicate==nil` and `errNoIndices`.
`BulkOp==true` is not supported and should never happen for a create operation.
If we don't panic or return a specific error here, we will be masking the usage problem.

File: go-controller/pkg/libovsdbops/model_client.go [hunk 6]
```diff
@@ -149,6 +150,8 @@ type operationModel struct {
 	BulkOp bool
 	// DoAfter is invoked at the end of the operation and allows to setup a
 	// subsequent operation with values obtained from this one.
+	// If model lookup was successful, Model will have UUID set,
```
Comment (jcaamano) [inline go-controller/pkg/libovsdbops/model_client.go:153]: Also true for create, not only lookups.

File: go-controller/pkg/ovn/pods.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (npinaeva) [reply go-controller/pkg/ovn/pods.go]: Yeah you're right, I guess this doesn't add any clarity (I just tried to make things more clear for those who read the code)

File: go-controller/pkg/libovsdbops/model_client.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (npinaeva) [reply go-controller/pkg/libovsdbops/model_client.go]: Ok I see, what we are trying to do is predict if user is trying to create a model or if he just wants to modify existing, and for the second option having no index and no predicate is an error. I would say there are some more unhandled ways to use the wrong combination of `operationModel` fields, but I added this exact check as `lookupRequired` function with some comments.

[the remark this answers is not in the document]
Comment (npinaeva) [reply go-controller/pkg/libovsdbops/model_client.go]: @jcaamano please check if that explanation makes sense, and if we can add something else

File: go-controller/pkg/libovsdbops/model_client.go [hunk 7]
```diff
@@ -385,53 +389,76 @@ func (m *modelClient) Lookup(opModels ...operationModel) error {
...
-		panic("Expected a ModelPredicate with BulkOp==true")
+		return m.whereCache(opModel)
 	}
-
-	return err
+	// the only operation that can be performed without a lookup (it can have no db indices and no ModelPredicate set)
+	// is Create.
+	if lookupRequired(opModel) {
+		return fmt.Errorf("provided operationModel should have either Model db index, or ModelPredicate set: " +
+			"lookup can't be performed")
```
Comment (jcaamano) [inline go-controller/pkg/libovsdbops/model_client.go:434]: I would rephrase to something like
`missing model indixes or predicate when a lookup was required`

File: go-controller/pkg/libovsdbops/model_client_test.go [hunk 8]
```diff
@@ -1318,6 +1313,37 @@ func TestLookup(t *testing.T) {
...
+			},
+		},
+		{
+			name: "Test lookup by predicate when bulk op",
+			op:   "Lookup",
+			generateOp: func() []operationModel {
+				return []operationModel{
+					{
+						Model:          &nbdb.AddressSet{},
+						ModelPredicate: func(item *nbdb.AddressSet) bool { return false },
```
[the remark this answers is not in the document]
Comment (npinaeva) [reply go-controller/pkg/libovsdbops/model_client_test.go:1333]: nice catch!

**Comment by jcaamano** [comment]
@dcbw do you have any remianing concerns in about comment https://github.com/ovn-org/ovn-kubernetes/pull/3142#issuecomment-1239514442 after my answer  https://github.com/ovn-org/ovn-kubernetes/pull/3142#issuecomment-1240619223?

**Comment by dcbw** [comment]
> @dcbw do you have any remianing concerns in about comment https://github.com/ovn-org/ovn-kubernetes/pull/3142#issuecomment-1239514442 after my answer https://github.com/ovn-org/ovn-kubernetes/pull/3142#issuecomment-1240619223?

@jcaamano I don't have any concerns. Thanks for the explanation.

**Comment by npinaeva** [comment]
this was causing unit test failure https://github.com/ovn-org/ovn-kubernetes/issues/3233
I found the same failure in other PRs, so this one shouldn't be the cause
I will post the fix for it in another PR, since it has nothing to do with model_client

**Review by jcaamano — APPROVED** [review]
